### PR TITLE
Form controls - `TextInput` component (02)

### PR DESCRIPTION
**Pull request by didoo**
### :pushpin: Summary

> **Note**
> This is not the final version: we will complete the component (documentation included) when the design specs are finalized.

This PR adds the `TextInput` form control to the system in two variants:
- `base`
- `field`

---

For the nomenclature see this illustration:
screenshot_1564

_Notice: the branch has been extracted from `form-controls/spike-playground` in #285 via a diff with the `main` branch and then cherry-picked only the changes needed for this PR._

### :hammer_and_wrench: Detailed description

In this PR we (me and @alex-ju) have:
- added the "base" and "field" `TextInput` form controls as components
- added the documentation page for the `TextInput`
- added integration tests and percy testing

Preview of the documentation page: https://hds-components-git-form-controls-02-text-input-hashicorp.vercel.app/components/form/text-input

Notice: the page is not linked in the index file of the dummy/scrappy website on purpose, so it's not visible to the public.

***

### 👀 How to review

👉 Review commit-by-commit or by files changed

Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

### CODE REVIEW

File: packages/components/app/styles/components/form/text-input.scss
Comment (alex-ju) [inline packages/components/app/styles/components/form/text-input.scss:97]: Thought: I've just realised we have this 'icon size' value in so many places we might want to convert it into a variable, in this file at least.

File: packages/components/tests/integration/components/hds/form/text-input/base-test.js
Comment (Dhaulagiri) [inline packages/components/tests/integration/components/hds/form/text-input/base-test.js:78]: [nitpick] as a reviewer i was expecting this to be next to the other tests for `@type` and started writing up a comment there before finding it down here. I assume there as informal convention on how these are being grouped, but what would you think about moving the tests to be next to each other?

File: packages/components/tests/integration/components/hds/form/text-input/base-test.js
Comment (Dhaulagiri) [inline packages/components/tests/integration/components/hds/form/text-input/base-test.js:80]: [suggestion] in the future if we were to change the list of input types we'll have to update this error message which feels a little too tightly coupled. for the purpose of this test we care that an error is captured, not necessarily what the exact message is. what do you think?

File: packages/components/tests/integration/components/hds/form/text-input/field-test.js
Comment (Dhaulagiri) [inline packages/components/tests/integration/components/hds/form/text-input/field-test.js:28]: [question] in the base tests we test for invalid `@type` values but don't here, is that intentional?

Comment (Dhaulagiri) [reply packages/components/tests/integration/components/hds/form/text-input/field-test.js:28]: _sidenote: i don't have a good suggestion otherwise but it strikes me how we are effectively copying the exact same tests from the base file into this wrapper component's tests which makes sense since we want to make sure consumers of the wrapper component have things work but feels also redudnant in a way I can't quite put my finger on_

**Review by Dhaulagiri — COMMENTED** [review]
I defer to @MelSumner @alex-ju  for a more thorough review but left a few comments thoughts that came up in my review

File: packages/components/app/styles/components/form/text-input.scss
Comment (didoo) [reply packages/components/app/styles/components/form/text-input.scss:41]: @alex-ju see comment here: https://github.com/hashicorp/design-system/pull/392#discussion_r902415176. I see the same issue with the `<input type="text">` unfortunately :(

File: packages/components/app/styles/components/form/text-input.scss
Comment (didoo) [reply packages/components/app/styles/components/form/text-input.scss:97]: Yes, the idea is to have a set of component-level design tokens, specific for the `form controls`, once the design specs are finalized, so we can share them with DevDot too.

File: packages/components/tests/integration/components/hds/form/text-input/base-test.js
Comment (didoo) [reply packages/components/tests/integration/components/hds/form/text-input/base-test.js:78]: By convention, all the "assertions" are collected at the bottom of the test files. <oving them around would spread them in no specific order, and make them easy to spot.

File: packages/components/tests/integration/components/hds/form/text-input/base-test.js
Comment (didoo) [reply packages/components/tests/integration/components/hds/form/text-input/base-test.js:80]: Yes, but that's on purpose. I could import (maybe?) the list from the component JS, but let's imagine I make an error there (eg. deleting one type), then the test would automatically auto-fix itself, and the error go unnoticed. Makes sense my thinking?

File: packages/components/tests/integration/components/hds/form/text-input/field-test.js
Comment (didoo) [reply packages/components/tests/integration/components/hds/form/text-input/field-test.js:28]: Yes I agree with you, but I went with redundancy in this case because we want to test that the property is correctly drilled down from the `field` to the `base` control.

File: packages/components/tests/dummy/app/templates/components/form/text-input.hbs
Comment (didoo) [reply packages/components/tests/dummy/app/templates/components/form/text-input.hbs:282]: Continuing the conversation from this PR: https://github.com/hashicorp/design-system/pull/285/files#r896982292

---

As discussed during the code review [the other day], I am not convinced this is the way to go for the "showcase" components. I totally understand the idea to have accessible documentation pages, but the _show_case is intended mainly to _show_ the components, with a particular focus on the developers implementing the components (and Percy) plus as a bonus side-effect, show to the consumers what are the variants (and their combinations) of a component. 

The issue of devs copying&pasting the code from the showcase is minimal, because we don't expect them to copy the HTML, and (unless I'm missing something) is not possible to copy the code from the Ember inspector.

So the last problems remains having pages which will not pass an accessibility test. Here I can propose some alternatives:
- we hide the whole showcase from the accessibility tree (a bit drastic!)
- we hide the parent container/wrapper of the "base" components from the accessibility tree (more specific)
- we accept that not all the elements in the showcase will meet all the accessibility criteria, when considered in isolation
  - this is not something unusual: consider this MDN page: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input - as you can see the top example has a label associated with the `<input>` but the ones in the list of types they don't and they also don't have a `aria-label` associated, and I am not sure I would expect them to have it frankly (their purpose meant to showcase the component, not as real usable/interactive content).

Comment (didoo) [reply packages/components/tests/dummy/app/templates/components/form/text-input.hbs:282]: I did a quick test on other design systems websites that we use as reference, these are the results for similar pages:
screenshot_1572
screenshot_1571
screenshot_1570
screenshot_1569

Comment (MelSumner) [reply packages/components/tests/dummy/app/templates/components/form/text-input.hbs:282]: Let's forget that this is something that matters to me (an equal member of the team), and look over the other reasons why this should be done. 

It's still inconsistent from the rest of our documentation pages and our approach so far.

I'm inferring two things from your response: 
1. someone with a screen reader will never need to use our documentation
2. other design systems have non-conformant documentation (so it's fine)

I think the first is perhaps the reason that we have to have accessibility laws in place-  what I have asked for is:
1. a small thing to fix 
2. a thing that will make our page as consistent as our other pages are
3. a thing that demonstrates our accessibility statement: intent to provide at least WCAG conformance

Even if I were to accept the argument that "other design systems have conformance issues in their documentation too", then compare the number of critical issues, because that's what this is. 

What would it mean then that a person who needs a screen reader will never visit this page?
1. that we would never hire someone who needed a screen reader
2. that none of us will never need a screen reader
3. that no auditor will ever be using a screen reader in their audit
4. that no site visitor (maybe works on another design system and has come to ours to see how we do things) will ever be using a screen reader

I don't think any of those things are logically safe to assume. 

However, even if I were to accept that a person who needs a screen reader were to never access this documentation, I have to wonder why you're so resistant to this. All of our other pages are conformant. This clearly is a critical WCAG issue. 

I'll even submit a PR to fix it.

File: packages/components/tests/integration/components/hds/form/text-input/base-test.js
Comment (Dhaulagiri) [reply packages/components/tests/integration/components/hds/form/text-input/base-test.js:80]: I think my suggestion is more that asserting that _any_ error is thrown is sufficient and what the actual contents of the error message are is not particularly high value

File: packages/components/tests/integration/components/hds/form/text-input/field-test.js
Comment (Dhaulagiri) [reply packages/components/tests/integration/components/hds/form/text-input/field-test.js:28]: makes sense, but should we also test for the invalid types here as well or no?

Comment (didoo) [reply packages/components/tests/integration/components/hds/form/text-input/field-test.js:28]: The invalid `@type` is tested in the `base` component (file `tests/integration/components/hds/form/text-input/base-test.js`, line 78), the `field` is consuming the `base` as inner component.

File: packages/components/tests/integration/components/hds/form/text-input/base-test.js
Comment (didoo) [reply packages/components/tests/integration/components/hds/form/text-input/base-test.js:80]: Ahh. I don't think there was a different way to test assertion (there is almost zero literature on this). I am not sure how to test assertions with just the "throw" part. Do you have ideas/suggestions?

Comment (Dhaulagiri) [reply packages/components/tests/integration/components/hds/form/text-input/base-test.js:80]: I'd need to do some digging. I think it's safe to table my suggestion for now and revisit if/when this is a problem.

**Review by didoo — COMMENTED** [review]
@alex-ju I've pushed a couple of small changes. for me is good to go, when you want feel free to merge.

**Review by alex-ju — APPROVED** [review]

File: packages/components/tests/dummy/app/templates/components/form/text-input.hbs
Comment (Dhaulagiri) [reply packages/components/tests/dummy/app/templates/components/form/text-input.hbs:439]: I opened a ticket in our backlog to track this discussion

**Pull request merged**
